### PR TITLE
melpaBuild: allow nix unstable version format

### PR DESCRIPTION
**Pull request by jian-lin**
Before, an unstable version of melpaBuild has to be in MELPA format, which has some disadvantages:
- existing update scripts cannot be reused[1]
- it causes inconsistence.

This patch fixes that by allowing an unstable version to be in Nix format.

We heuristically detect if it is an unstable version in Nix format. If so, we convert it to MELPA format.

This does not bring evaluation performance overhead.  Here is my benchmark result using hyperfine-1.18.0 and nix-2.18.1.

Without this patch:

```
Benchmark 1: NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.melpaPackages --apply 'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))' --no-eval-cache >/dev/null
  Time (mean ± σ):      9.126 s ±  0.048 s    [User: 6.531 s, System: 0.997 s]
  Range (min … max):    9.056 s …  9.186 s    10 runs
```

With this patch applied:

```
Benchmark 1: NIXPKGS_ALLOW_BROKEN=1 nix eval --include nixpkgs=$PWD --file . emacs.pkgs.melpaPackages --apply 'pkgSet: map (drv: drv.drvPath) (builtins.filter (p: p.type or null == "derivation") (builtins.attrValues pkgSet))' --no-eval-cache >/dev/null
  Time (mean ± σ):      9.124 s ±  0.057 s    [User: 6.522 s, System: 1.011 s]
  Range (min … max):    9.040 s …  9.220 s    10 runs
```

[1]: https://github.com/NixOS/nixpkgs/blob/177c9eddadc6abbaf62a19c90cc27a9fcde08e2d/pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix#L93-L108

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
I believe this is an extra reason to merge https://github.com/NixOS/nixpkgs/pull/315949 :)

File: pkgs/build-support/emacs/melpa.nix [hunk 1]
```diff
@@ -71,6 +76,18 @@ genericBuild ({
 
   inherit packageBuild commit ename recipe;
 
+  melpaVersion =
+    let
+      parsed = lib.flip builtins.match version
+        "^.*-unstable-([[:digit:]]{4})-([[:digit:]]{2})-([[:digit:]]{2})$";
+      unstableVersionInNixFormat = parsed != null; # heuristics
```
Comment (AndersonTorres) [inline pkgs/build-support/emacs/melpa.nix:83]:
```suggestion
  melpaVersion =
    let
      # Matches <version>-unstable-YYYY-MM-DD format
      parsed = lib.flip builtins.match version
        "^.*-unstable-([[:digit:]]{4})-([[:digit:]]{2})-([[:digit:]]{2})$";
      unstableVersionInNixFormat = parsed != null;
```

File: pkgs/build-support/emacs/melpa.nix [hunk 2]
```diff
@@ -96,17 +113,17 @@ genericBuild ({
         -L "$NIX_BUILD_TOP/package-build" \
         -l "$melpa2nix" \
         -f melpa2nix-build-package \
-        $ename $version $commit
+        $ename $melpaVersion $commit
```
Comment (AndersonTorres) [inline pkgs/build-support/emacs/melpa.nix:116]: Does it really work? I am impressed!

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/build-support/emacs/melpa.nix (hunk 2 above)
Comment (jian-lin) [reply pkgs/build-support/emacs/melpa.nix:116]: Not sure why you comment at this line.  But yes, the whole PR works.  You can try it yourself :)

Comment (AndersonTorres) [reply pkgs/build-support/emacs/melpa.nix:117]: This was mostly a reaction of 50% skeptic plus 50% amazed.

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Why is this PR taking so long to land?

https://nixpk.gs/pr-tracker.html?pr=316726

**Comment by jian-lin** [comment]
The [current PR](https://github.com/NixOS/nixpkgs/pull/316075) to update staging-next takes a bit more time than usual.

**Pull request merged**